extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/secret/KsmClient.java

```diff
@@ -56,280 +45,87 @@
-            cachedRecordsByUsername.clear();
-
-            // Store all records, sorting each into host-based and login-based
-            // buckets
-            records.forEach(record -> {
-
-                // Store based on UID ...
-                cachedRecordsByUid.put(record.getRecordUid(), record);
-
-                // ... and hostname/address
-                String hostname = recordService.getHostname(record);
-                addRecordForHost(record, hostname);
-
-                // Store based on username ONLY if no hostname (will otherwise
-                // result in ambiguous entries for servers tied to identical
-                // accounts)
-                if (hostname == null)
-                    addRecordForLogin(record, recordService.getUsername(record));
-
-            });
-
-            // Cache has been refreshed
-            this.cacheTimestamp = System.currentTimeMillis();
-
-        }
-        finally {
-            cacheLock.writeLock().unlock();
-        }
-
-    }
-
-    /**
-     * Associates the given record with the given hostname. The hostname may be
-     * null. Both {@link #cachedRecordsByHost} and {@link #cachedAmbiguousHosts}
-     * are updated appropriately. The write lock of {@link #cacheLock} must
-     * already be acquired before invoking this function.
+     * @param ksmConfig
+     *     The base-64 encoded JSON KSM config blob associated with the cache entry.
+     *     If an associated entry does not already exist, it will be created using
+     *     this configuration.
      *
-     * @param record
-     *     The record to associate with the hosts in the given field.
+     * @return
+     *     A KSM cache for the provided KSM config if not already present in the
+     *     cache map, otherwise the existing cache entry.
      *
-     * @param hostname
-     *     The hostname/address that the given record should be associated
-     *     with. This may be null.
+     * @throws GuacamoleException
+     *     If an error occurs while creating the KSM cache.
      */
-    private void addRecordForHost(KeeperRecord record, String hostname) {
+    private KsmCache createCacheIfNeeded(@Nullable String ksmConfig)
+            throws GuacamoleException {
 
-        if (hostname == null)
-            return;
-
-        KeeperRecord existing = cachedRecordsByHost.putIfAbsent(hostname, record);
-        if (existing != null && record != existing)
-            cachedAmbiguousHosts.add(hostname);
+        // If a cache already exists for the provided config, use it
+        KsmCache ksmCache = ksmCacheMap.get(ksmConfig);
+        if (ksmCache != null)
+            return ksmCache;
 
+        // Create and store a new KSM cache instance for the provided KSM config blob
+        SecretsManagerOptions options = confService.getSecretsManagerOptions(ksmConfig);
+        ksmCache = ksmCacheFactory.create(options);
+        ksmCacheMap.put(ksmConfig, ksmCache);
+        return ksmCache;
```

Comment:
Is this threadsafe?

Reply:
uh, no.

Reply:
Switched to using a concurrent map, which also required some refactoring around the way that properties are handled. Let me know what you think.

Reply:
Looks good - even in the case that `get()` returns `null` for two interleaved calls and two `KsmClient` instances get created, only one of those instances will win, everything requesting an instance will end up with the same instance, and the constructor that gets arguably unnecessarily invoked is very low overhead.